### ISSUE TITLE
KT-19310: Intellij can't resolve inner types in plugin class in build.gradle.kts files

### ISSUE THREAD

**Issue by JLLeitschuh** [gradle/kotlin-dsl#417]
Example:
```
plugins {
    id("com.diffplug.gradle.spotless") version "3.4.0"
    java
}

spotless {
    java {
        // The below line shows up red in Intellij
        eclipse()
    }
}

/**
 * Retrieves or configures the [spotless][com.diffplug.gradle.spotless.SpotlessExtension] project extension.
 */
fun Project.`spotless`(configure: com.diffplug.gradle.spotless.SpotlessExtension.() -> Unit = {}) =
    extensions.getByName<com.diffplug.gradle.spotless.SpotlessExtension>("spotless").apply { configure() }
```

<img width="949" alt="screen shot 2017-07-11 at 2 22 55 pm" src="https://user-images.githubusercontent.com/1323708/28084638-3bbd0f52-6647-11e7-8ff7-8cf5a0fd958d.png">

This is the method Intellij is trying to resolve the return type for:
https://github.com/diffplug/spotless/blob/master/plugin-gradle/src/main/java/com/diffplug/gradle/spotless/JavaExtension.java#L106-L108

I've experienced similar problems with other symbols.

For example this:
```
import com.diffplug.spotless.FormatterFunc

plugins {
    id("com.diffplug.gradle.spotless") version "3.4.0"
    java
}

fun toFormatterFunction(func : (String) -> String) : FormatterFunc {
    // The below line shows up red in Intellij
    return FormatterFunc(func)
}

spotless {
    java {
        custom("randomCheck", toFormatterFunction { it })
    }
}

/**
 * Retrieves or configures the [spotless][com.diffplug.gradle.spotless.SpotlessExtension] project extension.
 */
fun Project.`spotless`(configure: com.diffplug.gradle.spotless.SpotlessExtension.() -> Unit = {}) =
    extensions.getByName<com.diffplug.gradle.spotless.SpotlessExtension>("spotless").apply { configure() }
```

I've attached an example project.
The project runs fine under gradle but Intellj thinks there is a problem.
[test_spotless.zip](https://github.com/gradle/kotlin-dsl/files/1139952/test_spotless.zip)

Kotlin plugin version: `Version: 1.1.3-release-IJ2017.1-2`
Gradle version: `4.0.1`

```
IntelliJ IDEA 2017.1.5
Build #IU-171.4694.70, built on July 4, 2017
Licensed to Plexxi, Inc. / Jonathan Leitschuh
Subscription is active until December 14, 2017
JRE: 1.8.0_112-release-736-b21 x86_64
JVM: OpenJDK 64-Bit Server VM by JetBrains s.r.o
Mac OS X 10.12.5
```

**Comment by Barteks2x** [comment]
I'm also getting the same issue. Intellij can't find inner class, while everything works fine in gradle.

The code in my case:
```kotlin
configure<NamedDomainObjectContainer<ReobfTaskFactory.ReobfTaskWrapper>> {
    create("shadowJar").apply {
        mappingType = ReobfMappingType.SEARGE
    }
}
```

I'm trying to use [this class](https://github.com/MinecraftForge/ForgeGradle/blob/effc52cd92bbb743537dc59fa712bb1f9e45d309/src/main/java/net/minecraftforge/gradle/user/ReobfTaskFactory.java#L86), except for me Intellij shows `UnresolvedReference: ReobfTaskWrapper`. Ths interesting part is that I can use autocomplete, type `ReobfTaskW` and then ask for autocomplete, it will the suggest `ReobfTaskWrapper`, and when I select it, I end up with the same erroring code (that works just fine, only Intellij complains)

**Comment by JLLeitschuh** [comment]
I opened this issue against the Kotlin YouTrack issue board. Link to issue here:
https://youtrack.jetbrains.com/issue/KT-19310

**Comment by JLLeitschuh** [comment]
This should be a 1.0.0 milestone bug. This will be a big show stopper for new developers coming in when 1.0 is released.

**Comment by bamboo** [comment]
Agreed.

**Comment by JLLeitschuh** [comment]
Supposedly KT-19310 has been resolved. I'm not sure what release the fix will be going into though.

**Comment by eskatos** [comment]
The upstream issue is now assigned to Kotlin 1.1.50 which included the fix.

**Comment by eskatos** [comment]
Fix confirmed using IntelliJ 2017.2.5 & Kotlin 1.1.50, closing
